File: Two Pointers/Medium/LET_no16_3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int res = INT_MAX, n = nums.size();
        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int left = i + 1, right = n - 1;
            while (left < right) {
                int cur = nums[left] + nums[right] + nums[i];
                if (abs(long(target) - cur) < abs(long(target) - res))
                    res = cur;

                if (cur < target) {
                    left++;
                    while (left < right && nums[left] == nums[left - 1])
                        left++;
                }
                else {
                    right--;
                    while (left < right && nums[right] == nums[right + 1])
                        right--;
                }
            }
        }

        return res;
    }
};

File: Two Pointers/Medium/LET_no16_3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        res = sum(nums[:3])
        nums.sort()
        n = len(nums)
        
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            left, right = i + 1, n - 1
            while left < right:
                cur = nums[i] + nums[left] + nums[right]
                if cur == target:
                    return cur
                
                if abs(target - cur) < abs(target - res):
                    res = cur
                if cur < target:
                    left += 1
                else:
                    right -= 1
        
        return res
